### PR TITLE
NIFI-6855 - added action type enforcement option for handlers

### PR DESCRIPTION
**Pull request by YolandaMDavis**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java [hunk 1]
```diff
@@ -16,22 +16,54 @@
...
+            .displayName("Enforce Action Type")
+            .required(false)
+            .description("The Action Type that should be supported by this handler.  If provided any other type an " +
+                    "exception will be thrown")
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public void execute(Action action, Map<String, Object> facts){
+        if(actionTypeNotSupported(action)){
+            throw new UnsupportedOperationException("This Action Handler does not support actions with the provided type: " + action.getType());
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java]: Should we just ignore the action if it is not supported? For example, using the test file `test_nifi_rules.yml` I have a LOG and ALERT actions. If I am not using a correctly-configured ActionHandlerLookup and instead using either a single LogHandler or AlertHandler, then when both rules fire, I will get an exception on the handler. I was picturing using this property to "suppress" or filter the fired rules based on the configured handler, so only the LOG rules would actually fire if I my action handler is a LogHandler with Enforce Types set to LOG. What do you think?

File: nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java [hunk 2]
```diff
@@ -16,22 +16,54 @@
...
     public static enum DebugLevels {
         trace, debug, info, warn, error
     }
-    public abstract void execute(Action action, Map<String, Object> facts);
+
+    public static final PropertyDescriptor ENFORCE_ACTION_TYPE = new PropertyDescriptor.Builder()
+            .name("action-handler-enforce-type")
+            .displayName("Enforce Action Type")
+            .required(false)
+            .description("The Action Type that should be supported by this handler.  If provided any other type an " +
```
[the remark this answers is not in the document]
Comment (YolandaMDavis) [reply nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java]: I think that's reasonable, if people want to use multiple types to apply to a handler

File: nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java (hunk 1 above)
Comment (YolandaMDavis) [reply nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java]: The original intent was for users to have a way to ensure that only one type is supported and to immediately see (and flag as an issue) that something else is being sent. But I think there is value in giving an option for either mode (ignore vs enforce).

File: nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattyb149) [reply nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java]: My thought was for when the person that writes the rules is different from the person that configures the handler, and that maybe the rules file will include everything they'd ever want, but only want to turn on certain handlers / types at certain times. I didn't wrap my head fully around the use case, but I figure the more flexibility and power the better :)

Speaking of which, I think we should support EL here too, perhaps on a dev machine they are using the same rules file but don't want to fire alerts or SEND events, where in production they would.

[the remark this answers is not in the document]
Comment (YolandaMDavis) [reply nifi-nar-bundles/nifi-rules-action-handler-bundle/nifi-rules-action-handler-service/src/main/java/org/apache/nifi/rules/handlers/AbstractActionHandlerService.java]: Sure I can add EL here as well

**Comment by YolandaMDavis** [comment]
@mattyb149 thanks for your feedback thus far.  I've posted changes for your review

**Comment by mattyb149** [comment]
+1 LGTM, tested each of the Enforce policies for various actions/handlers. Thanks for this improvement! Merging to master

**Pull request closed without merging**